### ISSUE TITLE
Inconsistent results in LASSO Regression - Chapter 3, Section 6

### ISSUE THREAD

**Issue by jimthompson5802** [nfmcclure/tensorflow_cookbook#65]
@nfmcclure ... while working on the LASSO regression I noticed one issue.  As with the other notebooks, I added code for reproducibility.  Depending on the random seed, the LASSO regression returns `nan` results. 

Here are some examples.

LASSO code fragment:
```
# For Lasso, uncomment the following four lines (and comment out the Ridge Regression loss below)
lasso_param = tf.constant(0.9)
heavyside_step = tf.truediv(1., tf.add(1., tf.exp(tf.multiply(-100., tf.subtract(A, lasso_param)))))
regularization_param = tf.multiply(heavyside_step, 99.)
loss = tf.add(tf.reduce_mean(tf.square(y_target - model_output)), regularization_param)
```

With this seed setting:
```
seed = 13
np.random.seed(seed)
tf.set_random_seed(seed)
```

The training code
```
# Initialize variables
init = tf.global_variables_initializer()
sess.run(init)

# Training loop
loss_vec = []
for i in range(1500):
    rand_index = np.random.choice(len(x_vals), size=batch_size)
    rand_x = np.transpose([x_vals[rand_index]])
    rand_y = np.transpose([y_vals[rand_index]])
    sess.run(train_step, feed_dict={x_data: rand_x, y_target: rand_y})
    temp_loss = sess.run(loss, feed_dict={x_data: rand_x, y_target: rand_y})
    loss_vec.append(temp_loss[0])
    if (i+1)%300==0:
        print('Step #' + str(i+1) + ' A = ' + str(sess.run(A)) + ' b = ' + str(sess.run(b)))
        print('Loss = ' + str(temp_loss))
        #print('regularization_param='+ str(regularization_param))
        #print('heavyside step=' + str(sess.run(heavyside_step)))
        #print('\n')
```
Results in this output:
```
Step #300 A = [[ nan]] b = [[ nan]]
Loss = [[ nan]]
Step #600 A = [[ nan]] b = [[ nan]]
Loss = [[ nan]]
Step #900 A = [[ nan]] b = [[ nan]]
Loss = [[ nan]]
Step #1200 A = [[ nan]] b = [[ nan]]
Loss = [[ nan]]
Step #1500 A = [[ nan]] b = [[ nan]]
Loss = [[ nan]]
```

With this random seed setting:
```
seed = 31
np.random.seed(seed)
tf.set_random_seed(seed)
```
Results in this output:
```
Step #300 A = [[ 0.82903057]] b = [[ 1.62271953]]
Loss = [[ 10.61604691]]
Step #600 A = [[ 0.82230437]] b = [[ 3.08032537]]
Loss = [[ 3.35721326]]
Step #900 A = [[ 0.81690294]] b = [[ 3.88035631]]
Loss = [[ 1.26691902]]
Step #1200 A = [[ 0.81232369]] b = [[ 4.32818317]]
Loss = [[ 0.69326025]]
Step #1500 A = [[ 0.80592763]] b = [[ 4.57499456]]
Loss = [[ 0.33951065]]
```

Additional note, there are no issues with the Ridge Regression.

**Comment by nfmcclure** [comment]
Thanks @jimthompson5802 ,  I suspect this has to do with the continuous approximation to the heavyside step, and more specifically, the constant of 100.  I'll look into  playing around with this equation and seeing if I can get it to converge for more cases..

**Comment by nfmcclure** [comment]
Yup.  Making the continous heavyside step constant closer to zero results in a less-steep approximation, which means the gradients during the training step have a less likely change to explode (which results in the NaN values).

Changing the heavyside step function constant from -100.0 to -50.0:

`heavyside_step = tf.truediv(1., tf.add(1., tf.exp(tf.multiply(-50., tf.subtract(A, lasso_param)))))`

helped the gradients.  I hesitate to reduce it too much, because then we are moving towards a ridge regression and not the strict boundary of the lasso regression.

Let me know if you have any questions and thanks for finding this!
